### ISSUE TITLE
Flow 0.72, '*' as error disables generic type usage with "*" meaning "Use generic type, subtype does not matter"

### ISSUE THREAD

**Issue by lll000111** [facebook/flow#6294]
**ORIGINAL TEXT** (obsolete, see UPDATE below)

First, I know it's listed in the changelog, but that is not "documentation". What does it mean that the `*` type, which meant "Flow - you figure this type out" AFAIK is deprecated? What am I to use, `any`? Isn't this worth an explanation??? I'm pretty sure that internally at FB you didn't just dump this on your developers, or did somebody just declare `*` as "deprecated" overnight and everybody else just nodded fervently, "of course it is!"?

Flow Blog has the last entry 16 March, so nothing in the blog either. Google writes such a nice blog entry for each new Chrome release with everything relevant for developers.

And why should I not use `*` anyway? I'm using it mostly as the subtype in some generics, when it does not matter which subtype it actually is. It makes no sense at all for me to write anything else but `* `  in those generic types, for example `ObjectCreation<*>`, because which subtypes are allowed has already been specified in the type definition of the generic type, and all of them are allowed in those places where I use `*` instead of a concrete subtype.

Apart from my own generics, to illustrate the problem using well-known generic types: What about arrays or promises, when I only care that it is an array or a promise but not about the contents?

<br>

**EDIT / UPDATE /NEW TEXT**

**Generics** are a special case &mdash; when I only care about the parent type. In that case specifying the allowed subtype is exactly the same as the generics type definition, which is quote a waste of horizontal space. The `*` allowed me to concentrate on what was important, the parent type. So even when it works I now have to specify the sometimes lengthy (union type!) sub type in generic types even when the subtype does not matter. Having `*` was much better not just for space, it also captured the intent much better, that the subtype does not matter.

I don't have a single `*` apart from generic subtypes, so in general I'm okay with this restriction because I already implemented it myself, setting a concrete type everywhere. I just think generics were not considered, or not fully, because the parent type may very well be the most concrete type.

For example, instead of `ObjectCreation<*>` I now have to write `ObjectCreation<OneVersionedObjectTypes | OneUnversionedObjectTypes |  OneApplicationTypes>` because that is how the type was defined, and in the places that thus far had an `*` for the subtype only the `ObjectCreation` parent type itself mattered. So now my line is full horizontally when I just write the type. I cannot use `mixed` there, I tried, because `mixed` is not the same as the object union that the generic type was defined with (and it would be wrong, because it isn't anything at all that is possible).

Another example is general generic types such as arrays or promises: When I have functions that only uses the parent type itself and does not care what's inside. For example, when I use a promise for synchronization only but don't care or use the value it resolves with. I may not even know what the subtype will be in such a case, nor does that matter!

<br>

**TL;DR: Needed is a way to refer to only the parent type of a generic type, disregarding the subtype.** Thus far one could just use `MyGenericType<*>` for that, now the `*` is deprecated altogether, IMO throwing out the baby with the bathwater.

<br>

**Comment by TrySound** [comment]
Use `mixed` if you need. Most of the cases may be handled with explicit annotations and implicit inference.

**Comment by sibelius** [comment]
it is on release notes https://github.com/facebook/flow/releases/tag/v0.72.0

and this commit https://github.com/facebook/flow/commit/39968bcd1f5badccfd4303b2245a88203fa49377

**Comment by lll000111** [comment]
@sibelius Did you read my comment?

@TrySound `mixed`is BAD!!! It forces me to do checks because now Flow thinks it could be anything!

**Comment by TrySound** [comment]
Then rely on inference. * makes flow slower.

**Comment by lll000111** [comment]
@TrySound Then how do I get rid of the errors? There is nothing in the changelog even. Hence my issue.

**Comment by sibelius** [comment]
I think flow needs something like Typescript `infer`

**Comment by lll000111** [comment]
**Generics** are a special case &mdash; when I only care about the parent type. In that case specifying the allowed subtype is exactly the same as the generics type definition, which is quote a waste of horizontal space. The `*` allowed me to concentrate on what was important, the parent type. So even when it works I now have to specify the sometimes lengthy (union type!) sub type in generic types even when the subtype does not matter. Having `*` was much better not just for space, it also captured the intent much better, that the subtype does not matter.

**Comment by oriSomething** [comment]
I don't want it to sound sarcastic but, if `*` will be removed which means a lot of non-basic infers will not be exists, Why would I rather use Flow over TS?
I think since TS support nullability checking the "inferring" part is all the "big advantage" that left to TS. If it will be only very basic, and also I can see there is a new TS like syntax for calling functions (`f<number>();`) It seems the gap between the two is become narrower in a way I'm sure why prefer one on the other.

**Comment by lll000111** [comment]
@oriSomething I think specifying no type is the alternative, as far as I know, which is not very well though, Flow then autodetects? But subtypes within generic types don't have that alternative.

**Comment by oriSomething** [comment]
@lll000111 I'm not sure I understand. Can you please give a code example?

**Comment by lll000111** [comment]
@oriSomething For what? Not specifying a type is "" and generics syntax is generics syntax.

**Comment by steida** [comment]
@lll000111 I think you can use it like this. Note Input and Response.

```js
// @flow
import * as React from 'react';
import { type GraphQLTaggedNode } from 'react-relay';

// mutation, input, this.handleCompleted, this.handleError
export type Commit<Input, Response> = GraphQLTaggedNode => void;

const withMutation = <Props: {}, Input, Response>(
  Component: React.ComponentType<Props>,
): React.ComponentType<
  $Diff<
    Props,
    {
      commit: Commit<Input, Response> | void,
      pending: boolean | void,
    },
  >,
> => {
  return (props: Props) => (
    <Component {...props} commitMutation={42} pending={false} />
  );
};

export default withMutation;
```

**Comment by lll000111** [comment]
@steida That is what I already mentioned? See my example `ObjectCreation<OneVersionedObjectTypes | OneUnversionedObjectTypes | OneApplicationTypes>`.

**Comment by rj254** [comment]
How would the [recompose](https://github.com/acdlite/recompose/tree/master/types) types work without `*`? Lots of times the `*` replaces a much more complex type, especially when chaining HOCs

**Comment by sibelius** [comment]
For Recompose you can use Generics and $Call to make this work

**Comment by TrySound** [comment]
@rj254 Just describe that type with `React.ComponentType` instead of `HOC` and flow will just work.

**Comment by rj254** [comment]
@sibelius not sure I follow. In the example provided on the recompose site (copied here as well)

```
const enhance:HOC<*, EnhancedComponentProps> = compose(
  defaultProps({
    text: 'world',
  }),
  withProps(({ text }) => ({
    text: `Hello ${text}`
  }))
);
```

@TrySound I guess I could describe the enhanced component with `React.ComponentType`; however, my question was more geared to typing the `enhance` function.

**Comment by sibelius** [comment]
if you wanna get the return type of `enhance`

you can use this helper

```jsx
export type ExtractReturn<Fn> = $Call<<T>((...Iterable<any>) => T) => T, Fn>;

type EnhanceProps = ExtractReturn<enhance>
```

**Comment by lll000111** [comment]
Ping. Still very valid. Reminder: The issue is GENERIC TYPES, where the "*" still makes sense!

**Comment by girvo** [comment]
It's been a while since I worked on my current code-base, and I've been bitten by this. Specifically, for Lens generation: `Lens<S, A>` where I can specify `S` but want `A` to be inferred. It killed the library I used, too, it seems. This is such a shame, Flow was always my preferred tool for this because of it's powerful inference -- and being able to guide it, as such.

**Comment by mgtitimoli** [comment]
@girvo could you can try if `Lens<S, _>` works on your example (`_` can be used in places where we want that generic to be inferred)?
